In [1]:
import numpy as np
import pandas as pd
import geopandas as gpd
import requests
import pygeos
import osmnx as ox
import shapely as shp
from ast import literal_eval
from pyproj import Geod
import matplotlib.pyplot as plt
import folium

C:\Users\yianz\AppData\Local\Temp\ipykernel_3184\3262555724.py:3: UserWarning: Shapely 2.0 is installed, but because PyGEOS is also installed, GeoPandas will still use PyGEOS by default for now. To force to use and test Shapely 2.0, you have to set the environment variable USE_PYGEOS=0. You can do this before starting the Python process, or in your code before importing geopandas:

import os
os.environ['USE_PYGEOS'] = '0'
import geopandas

In a future release, GeoPandas will switch to using Shapely by default. If you are using PyGEOS directly (calling PyGEOS functions on geometries from GeoPandas), this will then stop working and you are encouraged to migrate from PyGEOS to Shapely 2.0 (https://shapely.readthedocs.io/en/latest/migration_pygeos.html).
  import geopandas as gpd


In [2]:
# Census tracts in NYC 
tracts = gpd.read_file('./data/new_york_census_tract.geojson')
tracts['GEOID_County'] = tracts.apply(lambda x:x['GEOID'][0:5], axis=1)
tracts.head(3)

GEOID  area_sqkm                                           geometry  \
0  36081003400   0.213985  POLYGON ((311370.732 57391.499, 311384.757 573...   
1  36081003900   0.146969  POLYGON ((304857.478 66110.315, 304948.710 663...   
2  36081003600   0.180006  POLYGON ((311522.128 57288.347, 311596.477 573...   

  GEOID_County  
0        36081  
1        36081  
2        36081

In [3]:
# Provider data 
df = pd.read_csv("data/Institutional_Provider_Network_Data__2021_Quarter_1.csv")
df = gpd.GeoDataFrame(df, geometry=gpd.points_from_xy(df['Longitude'], df['Latitude'], crs="EPSG:4326"))
df['Designated Service'] = df['Designated Service'].apply(lambda x: '{0:0>3}'.format(x))
df = df.to_crs(tracts.crs)

df.head(3)

C:\Users\yianz\AppData\Local\Temp\ipykernel_3184\1831518456.py:2: DtypeWarning: Columns (14,15,16,80) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("data/Institutional_Provider_Network_Data__2021_Quarter_1.csv")


Site Name Room/Suite  \
0                                       02 SOLUTIONS        NaN   
1                                       02 SOLUTIONS        NaN   
2  1 BETHESDA DRIVE OPERATING COMPANY LLC DB ELDE...        NaN   

         Address     Town/City  CountyFIPS County Name State  Zip Code  \
0  1406B ROUTE 9  CLIFTON PARK     36091.0    Saratoga    NY   12065.0   
1  1406B ROUTE 9  CLIFTON PARK     36091.0    Saratoga    NY   12065.0   
2  1 BETHESDA DR       HORNELL     36101.0     Steuben    NY   14843.0   

    Latitude  Longitude  ...  Servicing County 3  Servicing County 4  \
0  42.825080 -73.733701  ...                  35                  39   
1  42.825080 -73.733701  ...                 113                 115   
2  42.346198 -77.665526  ...                 999                 999   

  Servicing County 5  OASAS Program Reporting Unit Number  \
0                 57                                  NaN   
1                999                                  NaN   
2                999                                  NaN   

  OASAS Provider Number OMH ID                                Plan Name  \
0                   NaN    NaN                 Nascentia Health Options   
1                   NaN    NaN                 Nascentia Health Options   
2                   NaN    NaN  New York Quality Healthcare Corporation   

   Provider Network Data System Identification Number    Quarter/Year  \
0                                                 91   Quarter 1 2021   
1                                                 91   Quarter 1 2021   
2                                                 44   Quarter 1 2021   

                        geometry  
0  POINT (321788.653 295329.915)  
1  POINT (321788.653 295329.915)  
2   POINT (-2053.454 248396.682)  

[3 rows x 85 columns]

In [4]:
# Provider wiithin NYC
tract_geom_union = pygeos.from_shapely(tracts['geometry'].unary_union)

provider_geom = pygeos.from_shapely(df['geometry'].values)    
provider_geom_tree = pygeos.STRtree(provider_geom, leafsize=50)

query_intersect = provider_geom_tree.query(tract_geom_union, predicate='contains')
df_nyc = df.loc[query_intersect]
df_nyc.head(3)

Site Name Room/Suite            Address  \
1029    210 MJD PRECISION URGENT CARE        NaN  1737 SEAGIRT BLVD   
463720    SOUTH ISLAND MED ASSOCIATES        NaN  1735 SEAGIRT BLVD   
329494          NEW HEALTH ASSOCIATES        NaN  1731 SEAGIRT BLVD   

           Town/City  CountyFIPS County Name State  Zip Code   Latitude  \
1029    FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   
463720  FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   
329494  FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   

        Longitude  ...  Servicing County 3  Servicing County 4  \
1029   -73.751462  ...                 999                 999   
463720 -73.751460  ...                 999                 999   
329494 -73.751456  ...                 999                 999   

       Servicing County 5  OASAS Program Reporting Unit Number  \
1029                  999                                  NaN   
463720                999                                  NaN   
329494                999                                  NaN   

       OASAS Provider Number OMH ID  \
1029                     NaN    NaN   
463720                   NaN    NaN   
329494                   NaN    NaN   

                                                Plan Name  \
1029                 Consolidated Health Plans: MagnaCare   
463720                 UnitedHealthcare of New York, Inc.   
329494  Nippon Life Insurance Company of America: Mult...   

        Provider Network Data System Identification Number    Quarter/Year  \
1029                                                  169   Quarter 1 2021   
463720                                                148   Quarter 1 2021   
329494                                                111   Quarter 1 2021   

                            geometry  
1029    POINT (321038.397 47583.142)  
463720  POINT (321038.567 47583.143)  
329494  POINT (321038.905 47583.144)  

[3 rows x 85 columns]

In [5]:
df_nyc.columns.to_list()

['Site Name',
 'Room/Suite',
 'Address',
 'Town/City',
 'CountyFIPS',
 'County Name',
 'State',
 'Zip Code',
 'Latitude',
 'Longitude',
 'Area code',
 'Phone number',
 'Designated Service',
 'NPI',
 'License Number/Facility Operating Certificate',
 'PFI',
 'Medicaid Provider Identification Number',
 'Commercial Provider Indicator',
 'Medicaid Provider Indicator',
 'CHP Indicator',
 'HARP Indicator',
 'Medicaid Advantage Indicator',
 'Partial CAPS Indicator',
 'MAP Indicator',
 'PACE Indicator',
 'FIDA Indicator',
 'FIDAIDD Indicator',
 'HIVSNP Indicator=',
 'SIP-PL Indicator',
 'NYSOH Standard Essential Health Plan (EP) Indicator/Basic Health Program Indicator',
 'NYSOH EP Plus Adult Vision/Dental Indicator/Basic Health Program Adult Indicator',
 'Commercial Non-MCO Medical Indicator 1',
 'Commercial Non-MCO Medical Indicator 2',
 'Commercial Non-MCO Medical Indicator 3',
 'Commercial Non-MCO Medical Indicator 4',
 'Commercial Non-MCO Medical Indicator 5',
 'Commercial Non-MCO Medical 

In [6]:
urgent_desser = [149, 250, 901, 823]

In [7]:
df_urgent = df_nyc[df_nyc["Designated Service"].astype(int).isin(urgent_desser)]
df_urgent

Site Name Room/Suite  \
1027    210 MJD MEDICAL PLLC DBA SLVR STR URGENT        NaN   
453524                   SELVER STAR URGENT CARE        NaN   
1024         210 MJD MED PLLC DBA SILVER STAR UC        NaN   
323047                                       NaN        NaN   
458948                   SILVER STAR URGENT CARE        NaN   
...                                          ...        ...   
98748                         CITYMD URGENT CARE        NaN   
98752                         CITYMD URGENT CARE        NaN   
561280             WESTERN NY IMMEDIATE MED CARE        NaN   
561279             WESTERN NY IMMEDIATE MED CARE        NaN   
561278             WESTERN NY IMMEDIATE MED CARE        NaN   

                  Address     Town/City  CountyFIPS County Name State  \
1027    228 BEACH 20TH ST  FAR ROCKAWAY     36081.0      Queens    NY   
453524  228 BEACH 20TH ST  FAR ROCKAWAY     36081.0      Queens    NY   
1024    228 BEACH 20TH ST  FAR ROCKAWAY     36081.0      Queens    NY   
323047  228 BEACH 20TH ST  FAR ROCKAWAY     36081.0      Queens    NY   
458948  228 BEACH 20TH ST  FAR ROCKAWAY     36081.0      Queens    NY   
...                   ...           ...         ...         ...   ...   
98748     3494 JEROME AVE         BRONX     36005.0       Bronx    NY   
98752     3494 JEROME AVE         BRONX     36005.0       Bronx    NY   
561280    4256 BRONX BLVD         BRONX     36005.0       Bronx    NY   
561279    4256 BRONX BLVD         BRONX     36005.0       Bronx    NY   
561278    4256 BRONX BLVD         BRONX     36005.0       Bronx    NY   

        Zip Code   Latitude  Longitude  ...  Servicing County 3  \
1027     11691.0  40.596215 -73.753953  ...                 999   
453524   11691.0  40.596215 -73.753953  ...                 999   
1024     11691.0  40.596215 -73.753953  ...                 999   
323047   11691.0  40.596215 -73.753953  ...                 999   
458948   11691.0  40.596215 -73.753953  ...                 999   
...          ...        ...        ...  ...                 ...   
98748    10467.0  40.882527 -73.881181  ...                 999   
98752    10467.0  40.882527 -73.881181  ...                 999   
561280   10466.0  40.896509 -73.859827  ...                 999   
561279   10466.0  40.896509 -73.859827  ...                 999   
561278   10466.0  40.896509 -73.859827  ...                 999   

        Servicing County 4 Servicing County 5  \
1027                   999                999   
453524                 999                999   
1024                   999                999   
323047                 999                999   
458948                 999                999   
...                    ...                ...   
98748                  999                999   
98752                  999                999   
561280                 999                999   
561279                 999                999   
561278                 999                999   

        OASAS Program Reporting Unit Number OASAS Provider Number OMH ID  \
1027                                    NaN                   NaN    NaN   
453524                                  NaN                   NaN    NaN   
1024                                    NaN                   NaN    NaN   
323047                                  NaN                   NaN    NaN   
458948                                  NaN                   NaN    NaN   
...                                     ...                   ...    ...   
98748                                   NaN                   NaN    NaN   
98752                                   NaN                   NaN    NaN   
561280                                  NaN                   NaN    NaN   
561279                                  NaN                   NaN    NaN   
561278                                  NaN                   NaN    NaN   

                                                Plan Name  \
1027                                 

In [8]:
df_urgent_cleaned = df_urgent.groupby(["NPI", "Site Name"]).first().reset_index()
df_urgent_cleaned

NPI                                    Site Name Room/Suite  \
0    1003253030                             CURE URGENT CARE       None   
1    1003253030                            QUICK MED MEDICAL       None   
2    1003253030        QUICKMED MEDICAL DBA CURE URGENT CARE       None   
3    1003253030                          QUICKMED MEDICAL PC       None   
4    1003253030  QUICKMED MEDICAL, P.C. DBA CURE URGENT CARE       None   
..          ...                                          ...        ...   
830  9999999999              DOCTOR URGENT MEDICAL CARE PLLC       None   
831  9999999999                    FOREST URGENT CARE CENTER       None   
832  9999999999                    FORT HAMILTON URGENT CARE       None   
833  9999999999                               METRO URGICARE      Ste A   
834  9999999999               MJD MEDICAL PLLC DBA SILVER ST       None   

                Address      Town/City  CountyFIPS County Name State  \
0      2718 MERMAID AVE       BROOKLYN     36047.0       Kings    NY   
1         2689 BROADWAY       NEW YORK     36061.0    New York    NY   
2         175 E 96TH ST       NEW YORK     36061.0    New York    NY   
3      2718 MERMAID AVE       BROOKLYN     36047.0       Kings    NY   
4         175 E 96TH ST       NEW YORK     36061.0    New York    NY   
..                  ...            ...         ...         ...   ...   
830      44 SHERMAN AVE       NEW YORK     36061.0    New York    NY   
831      11201 75TH AVE   FOREST HILLS     36081.0      Queens    NY   
832       10031 4TH AVE       BROOKLYN     36047.0       Kings    NY   
833      542 E 138TH ST          BRONX     36005.0       Bronx    NY   
834  230 BEACH 102ND ST  ROCKAWAY PARK     36081.0      Queens    NY   

     Zip Code   Latitude  ...  Servicing County 3  Servicing County 4  \
0     11224.0  40.575491  ...                 999                 999   
1     10025.0  40.798872  ...                 999                 999   
2     10128.0  40.785570  ...                 999                 999   
3     11224.0  40.575491  ...                 999                 999   
4     10128.0  40.785570  ...                 999                 999   
..        ...        ...  ...                 ...                 ...   
830   10040.0  40.863165  ...                 999                 999   
831   11375.0  40.719156  ...                 999                 999   
832   11209.0  40.612113  ...                 999                 999   
833   10454.0  40.806932  ...                 999                 999   
834   11694.0  40.583932  ...                 999                 999   

     Servicing County 5 OASAS Program Reporting Unit Number  \
0                   999                                 NaN   
1                   999                                 NaN   
2                   999                                 NaN   
3                   999                                 NaN   
4                   999                                 NaN   
..                  ...                                 ...   
830                 999                                 NaN   
831                 999                                 NaN   
832                 999                                 NaN   
833                 999                                 NaN   
834                 999                                 NaN   

    OASAS Provider Number OMH ID  \
0                     NaN   None   
1                     NaN   None   
2                     NaN   None   
3                     NaN   None   
4                     NaN   None   
..                    ...    ...   
830                   NaN   None   
831                   NaN   None   
832                   NaN   None   
833                   NaN   None   
834                   NaN   None   

                                             Plan Name  \
0    Delaware American Life Insurance Company: Aetn...   
1    Healthfirst Health Plan, Inc. / Healthfirst PH...   
2    

In [9]:
def get_type(npi):
    # if npi is 888888888, it is exempt; if it is 9999999999, they have not yet received an npi
    if npi == 8888888888 or npi == 9999999999:
        return "na - exempt or not present"
    npi_api = f"https://npiregistry.cms.hhs.gov/api/?number={npi}&version=2.1"
    r = requests.get(npi_api)
    if not r.ok:
        return "error"
    try:
        return r.json()["results"][0]["enumeration_type"]
    except IndexError:
        return "na - no record found"

In [10]:
ny_npi = pd.read_csv("data/ny_npi.csv").set_index("NPI")
ny_npi["is_org"] = np.where(ny_npi["Provider Organization Name (Legal Business Name)"].isna(), "NPI-1", "NPI-2")

C:\Users\yianz\AppData\Local\Temp\ipykernel_3184\2536464180.py:1: DtypeWarning: Columns (3,4,5,6,7,8,9,10,11,13,14,15,16,17,18,24,32,39,40,41,42,43,44,45,60,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,115,118,119,121,123,125,127,129,130,131,133,134,135,137,138,139,141,142,143,145,146,147,149,150,151,153,154,155,157,158,159,161,162,163,165,166,167,169,170,171,173,174,175,177,178,179,181,182,183,185,186,187,189,190,191,193,194,195,197,198,199,201,202,203,205,206,207,209,210,211,213,214,215,217,218,219,221,222,223,225,226,227,229,230,231,233,234,235,237,238,239,241,242,243,245,246,247,249,250,251,253,254,255,257,258,259,261,262,263,265,266,267,269,270,271,273,274,275,277,278,279,281,283,285,286,287,289,290,291,293,294,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329) have mixed types. Specify dtype option on import or set low_memory=False.
  ny_npi = pd

In [11]:
def get_type_local(npi):
    # if npi is 888888888, it is exempt; if it is 9999999999, they have not yet received an npi
    if npi == 8888888888 or npi == 9999999999:
        return "na - exempt or not present"
    try:
        return ny_npi.loc[npi]["is_org"]
    except KeyError:
        return get_type(npi)

In [12]:
# df_urgent["type"] = df_urgent["NPI"].apply(lambda x: get_type_local(x))
df_urgent_cleaned["type"] = df_urgent_cleaned["NPI"].apply(lambda x: get_type_local(x))

In [13]:
# all records are organizations
df_urgent_cleaned["type"].value_counts()

NPI-2                         821
na - exempt or not present     14
Name: type, dtype: int64

In [14]:
# 823 is urgent care, 250 is emergency medicine, 149 is pediatric emergency medicine
df_urgent["Designated Service"].value_counts()

823    4627
250    1777
149       1
Name: Designated Service, dtype: int64

In [15]:
df_urgent.columns.to_list()

['Site Name',
 'Room/Suite',
 'Address',
 'Town/City',
 'CountyFIPS',
 'County Name',
 'State',
 'Zip Code',
 'Latitude',
 'Longitude',
 'Area code',
 'Phone number',
 'Designated Service',
 'NPI',
 'License Number/Facility Operating Certificate',
 'PFI',
 'Medicaid Provider Identification Number',
 'Commercial Provider Indicator',
 'Medicaid Provider Indicator',
 'CHP Indicator',
 'HARP Indicator',
 'Medicaid Advantage Indicator',
 'Partial CAPS Indicator',
 'MAP Indicator',
 'PACE Indicator',
 'FIDA Indicator',
 'FIDAIDD Indicator',
 'HIVSNP Indicator=',
 'SIP-PL Indicator',
 'NYSOH Standard Essential Health Plan (EP) Indicator/Basic Health Program Indicator',
 'NYSOH EP Plus Adult Vision/Dental Indicator/Basic Health Program Adult Indicator',
 'Commercial Non-MCO Medical Indicator 1',
 'Commercial Non-MCO Medical Indicator 2',
 'Commercial Non-MCO Medical Indicator 3',
 'Commercial Non-MCO Medical Indicator 4',
 'Commercial Non-MCO Medical Indicator 5',
 'Commercial Non-MCO Medical 

In [37]:
def construct_address(row):
    res = ""
    res += row["Address"]
    if (row["Room/Suite"] != None):
        res += " " + row["Room/Suite"]
    res += ", "
    res += row["Town/City"] + ", "
    res += row["State"]
    res += " " + str(int(row["Zip Code"]))
    return res

In [38]:
df_urgent_cleaned["full address"] = df_urgent_cleaned.apply(lambda x: construct_address(x), axis=1)

In [41]:
final_df = df_urgent_cleaned[["Site Name", "Address", "NPI"]]
final_df

Site Name             Address  \
0                               CURE URGENT CARE    2718 MERMAID AVE   
1                              QUICK MED MEDICAL       2689 BROADWAY   
2          QUICKMED MEDICAL DBA CURE URGENT CARE       175 E 96TH ST   
3                            QUICKMED MEDICAL PC    2718 MERMAID AVE   
4    QUICKMED MEDICAL, P.C. DBA CURE URGENT CARE       175 E 96TH ST   
..                                           ...                 ...   
830              DOCTOR URGENT MEDICAL CARE PLLC      44 SHERMAN AVE   
831                    FOREST URGENT CARE CENTER      11201 75TH AVE   
832                    FORT HAMILTON URGENT CARE       10031 4TH AVE   
833                               METRO URGICARE      542 E 138TH ST   
834               MJD MEDICAL PLLC DBA SILVER ST  230 BEACH 102ND ST   

            NPI  
0    1003253030  
1    1003253030  
2    1003253030  
3    1003253030  
4    1003253030  
..          ...  
830  9999999999  
831  9999999999  
832  9999999999  
833  9999999999  
834  9999999999  

[835 rows x 3 columns]

## Notes



benefits over health facility genearl information:

- more complete (e.g., has urgent care)
- have more in-depth segmentation (instead of clinic, have type of clinic and what it actually does, which can inform whether a facility supports walk-ins), (desser + service types)
    - e.g., hospital extension clinics in hfgi include a methadone center, a dance injury center, and a dentistry center while in pnds, they are segmented into more identifying categories
- includes the number of providers (noprov)

cons versus health facility general information

- more complex (has a bunch of random, irrelevant stuff)
- indexed by health facilities & plans based on which plans contract with what facilities -- not directly comparable to healthcare accessibility without some tinkering

documentation: https://www.health.ny.gov/health_care/managed_care/docs/dictionary.pdf

what columns mean: https://www.opendatanetwork.com/dataset/health.data.ny.gov/eary-k6ia

To use this data, we need to assume the following:

1. The institutions that are involved in group medical insurance plans are representative of all relevant institutions.
    - this is somewhat reasonable since these institutions would be the most accessible to the average person
2. The assigned designated service is representative of what a facility actually does and can be used to determine whether a facility supports walk-ins or not
    - (can't guarantee that walk-ins are supported but can make generalizations like that clinics generally allow walk-ins)
3. Some subset of the data can be validly excluded, as it is not relevant to medical accessibility
    - should ideally exclude all medically tangential businesses while including relatively relevant ones (like urgent care)

Potential service codes to look for (that might support walk-ins)

- 005 state operated clinic
- 011 general hospital
- 034 state operated facility
- 150 pediatrics
- 252 primary care initiative in underserved areas
- 760 pharmacy?
- 776 general mpractice only - nospec?
- 777 all physician?
- 823 urgent care
- 899 hospital inpatient?
- 901 emergency room
- 914 general medicine - clinic speciality ?
- 978 preferred primary care clinic?
- clinic specialty things (have basically one for every field)